// n8n-mcp/dist/utils/version-utils.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseVersion = parseVersion;
exports.compareVersions = compareVersions;
exports.getSupportedN8nVersion = getSupportedN8nVersion;
exports.getVersionCompatibilityStatus = getVersionCompatibilityStatus;
exports.formatVersion = formatVersion;
const fs = __importStar(require("fs"));
const path = __importStar(require("path"));
const logger_1 = require("./logger");
function parseVersion(version) {
    const cleanVersion = version.replace(/^[v^~]/, '');
    const match = cleanVersion.match(/^(\d+)\.(\d+)\.(\d+)/);
    if (!match) {
        logger_1.logger.warn(`Failed to parse version: ${version}`);
        return null;
    }
    return {
        major: parseInt(match[1], 10),
        minor: parseInt(match[2], 10),
        patch: parseInt(match[3], 10)
    };
}
function compareVersions(v1, v2) {
    if (v1.major !== v2.major) {
        return v1.major > v2.major ? 1 : -1;
    }
    if (v1.minor !== v2.minor) {
        return v1.minor > v2.minor ? 1 : -1;
    }
    if (v1.patch !== v2.patch) {
        return v1.patch > v2.patch ? 1 : -1;
    }
    return 0;
}
let cachedSupportedVersion = null;
function getSupportedN8nVersion() {
    if (cachedSupportedVersion) {
        return cachedSupportedVersion;
    }
    try {
        const packageJsonPath = path.join(__dirname, '..', '..', 'package.json');
        const packageJson = JSON.parse(fs.readFileSync(packageJsonPath, 'utf8'));
        const n8nVersion = packageJson.dependencies?.n8n;
        if (!n8nVersion) {
            logger_1.logger.warn('n8n dependency not found in package.json');
            return null;
        }
        const cleanedVersion = n8nVersion.replace(/^[^0-9]*/, '');
        cachedSupportedVersion = cleanedVersion;
        return cleanedVersion;
    }
    catch (error) {
        logger_1.logger.error('Failed to read supported n8n version from package.json', error);
        return null;
    }
}
function getVersionCompatibilityStatus(instanceVersion, supportedVersion) {
    if (!instanceVersion || !supportedVersion) {
        return {
            status: 'unknown',
            warning: 'Unable to determine version compatibility - version information missing'
        };
    }
    const instanceV = parseVersion(instanceVersion);
    const supportedV = parseVersion(supportedVersion);
    if (!instanceV || !supportedV) {
        return {
            status: 'unknown',
            warning: 'Unable to parse version numbers for comparison'
        };
    }
    const comparison = compareVersions(instanceV, supportedV);
    if (instanceV.major === supportedV.major && instanceV.minor === supportedV.minor) {
        return { status: 'compatible' };
    }
    if (comparison > 0) {
        return {
            status: 'outdated_mcp',
            warning: `Your n8n instance (${instanceVersion}) is newer than what this MCP version supports (${supportedVersion}). Some nodes may not work correctly. Please update n8n-mcp to the latest version.`
        };
    }
    if (comparison < 0) {
        return {
            status: 'outdated_n8n',
            warning: `Your n8n instance (${instanceVersion}) is older than what this MCP version supports (${supportedVersion}). Please update your n8n instance to version ${supportedVersion} or newer for best compatibility.`
        };
    }
    return { status: 'compatible' };
}
function formatVersion(version) {
    if (!version)
        return 'unknown';
    return version.startsWith('v') ? version : `v${version}`;
}
