// n8n-mcp/dist/templates/template-fetcher.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TemplateFetcher = void 0;
const axios_1 = __importDefault(require("axios"));
const logger_1 = require("../utils/logger");
class TemplateFetcher {
    constructor() {
        this.baseUrl = 'https://api.n8n.io/api/templates';
        this.pageSize = 250;
    }
    async fetchTemplates(progressCallback) {
        const allTemplates = await this.fetchAllTemplates(progressCallback);
        const oneYearAgo = new Date();
        oneYearAgo.setMonth(oneYearAgo.getMonth() - 12);
        const recentTemplates = allTemplates.filter((w) => {
            const createdDate = new Date(w.createdAt);
            return createdDate >= oneYearAgo;
        });
        logger_1.logger.info(`Filtered to ${recentTemplates.length} templates from last 12 months (out of ${allTemplates.length} total)`);
        return recentTemplates;
    }
    async fetchAllTemplates(progressCallback) {
        const allTemplates = [];
        let page = 1;
        let hasMore = true;
        let totalWorkflows = 0;
        logger_1.logger.info('Starting complete template fetch from n8n.io API');
        while (hasMore) {
            try {
                const response = await axios_1.default.get(`${this.baseUrl}/search`, {
                    params: {
                        page,
                        rows: this.pageSize
                    }
                });
                const { workflows } = response.data;
                totalWorkflows = response.data.totalWorkflows || totalWorkflows;
                allTemplates.push(...workflows);
                const totalPages = Math.ceil(totalWorkflows / this.pageSize);
                if (progressCallback) {
                    progressCallback(allTemplates.length, totalWorkflows);
                }
                logger_1.logger.debug(`Fetched page ${page}/${totalPages}: ${workflows.length} templates (total so far: ${allTemplates.length}/${totalWorkflows})`);
                if (workflows.length < this.pageSize) {
                    hasMore = false;
                }
                page++;
                if (hasMore) {
                    await this.sleep(300);
                }
            }
            catch (error) {
                logger_1.logger.error(`Error fetching templates page ${page}:`, error);
                throw error;
            }
        }
        logger_1.logger.info(`Fetched all ${allTemplates.length} templates from n8n.io`);
        return allTemplates;
    }
    async fetchTemplateDetail(workflowId) {
        try {
            const response = await axios_1.default.get(`${this.baseUrl}/workflows/${workflowId}`);
            return response.data.workflow;
        }
        catch (error) {
            logger_1.logger.error(`Error fetching template detail for ${workflowId}:`, error);
            throw error;
        }
    }
    async fetchAllTemplateDetails(workflows, progressCallback) {
        const details = new Map();
        logger_1.logger.info(`Fetching details for ${workflows.length} templates`);
        for (let i = 0; i < workflows.length; i++) {
            const workflow = workflows[i];
            try {
                const detail = await this.fetchTemplateDetail(workflow.id);
                details.set(workflow.id, detail);
                if (progressCallback) {
                    progressCallback(i + 1, workflows.length);
                }
                await this.sleep(150);
            }
            catch (error) {
                logger_1.logger.error(`Failed to fetch details for workflow ${workflow.id}:`, error);
            }
        }
        logger_1.logger.info(`Successfully fetched ${details.size} template details`);
        return details;
    }
    sleep(ms) {
        return new Promise(resolve => setTimeout(resolve, ms));
    }
}
exports.TemplateFetcher = TemplateFetcher;
