// n8n-mcp/dist/mcp/working-test.js
#!/usr/bin/env node
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const index_js_1 = require("@modelcontextprotocol/sdk/server/index.js");
const stdio_js_1 = require("@modelcontextprotocol/sdk/server/stdio.js");
console.error('[n8n-mcp] Starting working-test.ts');
const server = new index_js_1.Server({
    name: 'n8n-mcp',
    version: '2.7.5'
}, {
    capabilities: {
        tools: {}
    }
});
server.setRequestHandler('tools/list', async () => {
    console.error('[n8n-mcp] tools/list called');
    return {
        tools: [
            {
                name: 'test_tool',
                description: 'A test tool',
                inputSchema: {
                    type: 'object',
                    properties: {}
                }
            }
        ]
    };
});
async function main() {
    try {
        console.error('[n8n-mcp] Creating transport');
        const transport = new stdio_js_1.StdioServerTransport();
        console.error('[n8n-mcp] Connecting server');
        await server.connect(transport);
        console.error('[n8n-mcp] Server connected and running');
    }
    catch (error) {
        console.error('[n8n-mcp] Error:', error);
        process.exit(1);
    }
}
console.error('[n8n-mcp] About to start main()');
main().catch(error => {
    console.error('[n8n-mcp] Fatal error:', error);
    process.exit(1);
});
